### PR TITLE
Make cli code testable

### LINKED ISSUE #405: Add testing support for cli
Currently, our Stack setup in package.yaml has our project containing a library, an executable, and a test for the library. It would be good to extend this in some way that allows us to write tests for `cli` (our executable).

I have spent a few hours trying to modify package.yaml (read fight Stack + Cabal + hstack) such that we can add a new test stanza and ensure it comes out in the waspc.cabal correctly. This includes:

```
verbatim:
  build-tool-depends: waspc:wasp-cli
```

and 

```
build-tools:
  - waspc:wasp-cli
```

The crux of the issue is while I can get a correct looking cabal file, I still cannot import any `cli` modules from a test file in this new test suite. So something is not wiring up or working like I expected.

As a last ditch effort, I even tried to add `cli` to `source-dirs` and move over all of its `dependencies` to this new test stanza, but then I got a tasty-discover error.

Perhaps we just try alternative approaches?

Ref: Question I opened on Reddit that is awaiting moderator approval
https://www.reddit.com/r/haskell/comments/rh7km8/testing_an_executable_module_in_stack_project/

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

Restructures our project setup to leverage Cabal internal libraries so `cli` code can be shared by an executable and a test suite.

Before:
```
cli
├── Main.hs
└── Wasp
    └── Cli
```

After:
```
cli
├── exe
│   └── Main.hs
├── src
│   └── Wasp
│       └── Cli
└── test
    ├── TastyDiscoverDriver.hs
    └── TerminalTest.hs
```

Ref: https://cabal.readthedocs.io/en/latest/cabal-package.html#sublibs

It includes one simple example test to show it is possible to access CLI code to test.

Closes #405 

## Type of change

Please select the option(s) that is more relevant.

- [x] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

File: waspc/run [hunk 1]
```diff
@@ -114,6 +115,9 @@ case $COMMAND in
             echo_and_eval "$TEST_UNIT_CMD --test-options \"-p \\\"$TEST_PATTERN\\\"\""
         fi
 
+        ;;
+    test:cli)
```
Comment (Martinsos) [inline waspc/run:119]: You will also want to add `test-cli` above in the "usage" output of the command, so people can learn about it.

**Comment by Martinsos** [comment]
Nice job @shayneczyzewski !

I am wondering if we should structure it as `cli/exe`, `cli/src` and `cli/test`?

Or we could even go as far as to make it a separate package now -> a thing that has lib, exe and tests -> would it make any sense to make a separate package? Or would that just introduce too much complexity? I never did it so I don't know much.

We shouldn't do any of these things if they will waste a lot of time and don't bring anything, I am just wondering what is the nicest way to structure it, since there seem to be multiple options.

File: waspc/run (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/run:119]: Good catch, thanks will do

**Comment by shayneczyzewski** [comment]
> Nice job @shayneczyzewski !
> 
> 
> 
> I am wondering if we should structure it as `cli/exe`, `cli/src` and `cli/test`?
> 
> 
> 
> Or we could even go as far as to make it a separate package now -> a thing that has lib, exe and tests -> would it make any sense to make a separate package? Or would that just introduce too much complexity? I never did it so I don't know much.
> 
> 
> 
> We shouldn't do any of these things if they will waste a lot of time and don't bring anything, I am just wondering what is the nicest way to structure it, since there seem to be multiple options.

Thanks @Martinsos! I can certainly look into putting everything under `cli` tomorrow. The separate package may take longer/be overkill, but will at least give a quick research before deciding and report back. Appreciate the eyes on this!

**Comment by Martinsos** [comment]
> > Nice job @shayneczyzewski !
> > I am wondering if we should structure it as `cli/exe`, `cli/src` and `cli/test`?
> > Or we could even go as far as to make it a separate package now -> a thing that has lib, exe and tests -> would it make any sense to make a separate package? Or would that just introduce too much complexity? I never did it so I don't know much.
> > We shouldn't do any of these things if they will waste a lot of time and don't bring anything, I am just wondering what is the nicest way to structure it, since there seem to be multiple options.
> 
> Thanks @Martinsos! I can certainly look into putting everything under `cli` tomorrow. That should help tidy the dir structure. The separate package may take longer/be overkill, but will at least give a quick research before deciding and report back. Appreciate the eyes on this!

Sounds good -> yeah I agree we shouldn't overcomplicate now, if this works that is great and we can move on :). Just trying to get a sense for other options.

**Review by Martinsos — APPROVED** [review]

**Comment by shayneczyzewski** [comment]
Thanks for the final check and review @Martinsos! I think the new structure under `cli` looks much better, thanks for that! 👍🏻 I did some research on separate package but think that may be premature at this point (i.e. what do we gain/lose at this point, not sure I have enough experience with Cabal to know right now), but good to keep in mind. Thanks again 🙏🏻

**Pull request merged**